parlai/core/torch_generator_agent.py

```diff
@@ -0,0 +1,1022 @@
+
+            # generated response
+            scores = out[0]
+            _, preds = scores.max(2)
+
+            score_view = scores.view(-1, scores.size(-1))
+            loss = self.criterion(score_view, batch.label_vec.view(-1))
+            # save loss to metrics
+            notnull = batch.label_vec.ne(self.NULL_IDX)
+            target_tokens = notnull.long().sum().item()
+            correct = ((batch.label_vec == preds) * notnull).sum().item()
+            self.metrics['correct_tokens'] += correct
+            self.metrics['loss'] += loss.item()
+            self.metrics['num_tokens'] += target_tokens
+            loss /= target_tokens  # average loss per token
+            loss.backward()
+            self.update_params()
+        except RuntimeError as e:
+            # catch out of memory exceptions during fwd/bck (skip batch)
+            if 'out of memory' in str(e):
+                print('| WARNING: ran out of memory, skipping batch. '
+                      'if this happens frequently, decrease batchsize or '
+                      'truncate the inputs to the model.')
+                self.metrics['total_skipped_batches'] += 1
+            else:
+                raise e
+
+    def _pick_cands(self, cand_preds, cand_inds, cands):
+        cand_replies = [None] * len(cands)
+        for idx, order in enumerate(cand_preds):
+            batch_idx = cand_inds[idx]
+            cand_replies[batch_idx] = [cands[batch_idx][i] for i in order]
+        return cand_replies
+
+    def _write_beam_dots(self, text_vecs, beams):
+        """Write the beam dot files to disk."""
+        for i, b in enumerate(beams):
+            dot_graph = b.get_beam_dot(dictionary=self.dict, n_best=3)
+            image_name = self._v2t(text_vecs[i, -20:])
+            image_name = image_name.replace(' ', '-').replace('__null__', '')
+            dot_graph.write_png(
+                os.path.join(self.beam_dot_dir, "{}.png".format(image_name))
+            )
+
+    def eval_step(self, batch):
+        """Evaluate a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        bsz = batch.text_vec.size(0)
+        self.model.eval()
+        cand_scores = None
+
+        if self.skip_generation:
+            global _WARNING_SEEN
+            if not _WARNING_SEEN:
+                _WARNING_SEEN = True
+                import warnings
+                warnings.warn(
+                    "--skip-generation does not produce accurate metrics beyond ppl",
+                    RuntimeWarning
+                )
+            out = self.model(batch.text_vec, batch.label_vec)
+            scores = out[0]
+            _, preds = scores.max(2)
+        elif self.beam_size == 1:
+            preds, _ = self.model(batch.text_vec)
+        elif self.beam_size > 1:
+            out = self.beam_search(
+                self.model,
+                batch,
+                self.beam_size,
+                start=self.START_IDX,
```

Comment:
cleanup params here. Beam search is a class method so we don't need to pass some of these params

Reply:
@uralik and I will be doing iterative beam on top of this, so I'm actually going to hold off on this parameter cleanup until a later time.